File: scripts/test_pattern/test_factory_method.py
"""Tests for the Factory Method pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from factory_method import (
    RoadLogistics, SeaLogistics, AirLogistics,
    Truck, Ship, Plane, make_logistics
)


class TestFactoryMethod(unittest.TestCase):

    # ------------------------------------------------------------------
    # Product creation
    # ------------------------------------------------------------------
    def test_road_logistics_creates_truck(self):
        t = RoadLogistics().create_transport()
        self.assertIsInstance(t, Truck)
        self.assertEqual("Truck", t.type())

    def test_sea_logistics_creates_ship(self):
        t = SeaLogistics().create_transport()
        self.assertIsInstance(t, Ship)
        self.assertEqual("Ship", t.type())

    def test_air_logistics_creates_plane(self):
        t = AirLogistics().create_transport()
        self.assertIsInstance(t, Plane)
        self.assertEqual("Plane", t.type())

    # ------------------------------------------------------------------
    # Deliver messages
    # ------------------------------------------------------------------
    def test_truck_deliver_message(self):
        self.assertIn("land", Truck().deliver())

    def test_ship_deliver_message(self):
        self.assertIn("sea", Ship().deliver())

    def test_plane_deliver_message(self):
        self.assertIn("air", Plane().deliver())

    # ------------------------------------------------------------------
    # plan_delivery (template method using factory)
    # ------------------------------------------------------------------
    def test_plan_delivery_road(self):
        self.assertIn("Truck", RoadLogistics().plan_delivery())

    def test_plan_delivery_sea(self):
        self.assertIn("Ship", SeaLogistics().plan_delivery())

    def test_plan_delivery_air(self):
        self.assertIn("Plane", AirLogistics().plan_delivery())

    # ------------------------------------------------------------------
    # Helper factory function
    # ------------------------------------------------------------------
    def test_make_logistics_road(self):
        l = make_logistics("road")
        self.assertIsInstance(l, RoadLogistics)

    def test_make_logistics_sea(self):
        l = make_logistics("sea")
        self.assertIsInstance(l, SeaLogistics)

    def test_make_logistics_air(self):
        l = make_logistics("air")
        self.assertIsInstance(l, AirLogistics)

    def test_make_logistics_unknown_raises(self):
        with self.assertRaises(ValueError):
            make_logistics("teleport")

    # ------------------------------------------------------------------
    # Each creation returns a new instance
    # ------------------------------------------------------------------
    def test_each_creation_is_unique(self):
        logistics = RoadLogistics()
        t1 = logistics.create_transport()
        t2 = logistics.create_transport()
        self.assertIsNot(t1, t2)


if __name__ == "__main__":
    unittest.main()

File: test/pattern/test_factory_method.cpp
#include "pattern/factory_method.hpp"
#include <gtest/gtest.h>

using namespace pattern;

// ---------------------------------------------------------------------------
// Product creation tests
// ---------------------------------------------------------------------------
TEST(FactoryMethod, RoadLogisticsCreatesTruck)
{
    RoadLogistics logistics;
    auto          transport = logistics.createTransport();
    ASSERT_NE(nullptr, transport.get());
    EXPECT_EQ("Truck", transport->type());
}

TEST(FactoryMethod, SeaLogisticsCreatesShip)
{
    SeaLogistics logistics;
    auto         transport = logistics.createTransport();
    ASSERT_NE(nullptr, transport.get());
    EXPECT_EQ("Ship", transport->type());
}

TEST(FactoryMethod, AirLogisticsCreatesPlane)
{
    AirLogistics logistics;
    auto         transport = logistics.createTransport();
    ASSERT_NE(nullptr, transport.get());
    EXPECT_EQ("Plane", transport->type());
}

// ---------------------------------------------------------------------------
// Deliver message tests
// ---------------------------------------------------------------------------
TEST(FactoryMethod, TruckDeliverMessage)
{
    Truck t;
    EXPECT_NE(std::string::npos, t.deliver().find("land"));
}

TEST(FactoryMethod, ShipDeliverMessage)
{
    Ship s;
    EXPECT_NE(std::string::npos, s.deliver().find("sea"));
}

TEST(FactoryMethod, PlaneDeliverMessage)
{
    Plane p;
    EXPECT_NE(std::string::npos, p.deliver().find("air"));
}

// ---------------------------------------------------------------------------
// planDelivery (template method using factory method)
// ---------------------------------------------------------------------------
TEST(FactoryMethod, PlanDeliveryUsesCorrectTransport)
{
    RoadLogistics road;
    EXPECT_NE(std::string::npos, road.planDelivery().find("Truck"));

    SeaLogistics sea;
    EXPECT_NE(std::string::npos, sea.planDelivery().find("Ship"));

    AirLogistics air;
    EXPECT_NE(std::string::npos, air.planDelivery().find("Plane"));
}

// ---------------------------------------------------------------------------
// Polymorphism through base pointer
// ---------------------------------------------------------------------------
TEST(FactoryMethod, PolymorphicUsage)
{
    std::unique_ptr<Logistics> logistics = std::unique_ptr<Logistics>(new SeaLogistics());
    auto                       transport = logistics->createTransport();
    EXPECT_EQ("Ship", transport->type());
}

// ---------------------------------------------------------------------------
// Helper factory function
// ---------------------------------------------------------------------------
TEST(FactoryMethod, MakeLogisticsRoad)
{
    auto l = makeLogistics("road");
    ASSERT_NE(nullptr, l.get());
    EXPECT_NE(std::string::npos, l->planDelivery().find("Truck"));
}

TEST(FactoryMethod, MakeLogisticsUnknownThrows)
{
    EXPECT_THROW(makeLogistics("teleport"), std::invalid_argument);
}

// ---------------------------------------------------------------------------
// Each call to createTransport returns a distinct object (not shared)
// ---------------------------------------------------------------------------
TEST(FactoryMethod, EachCreationIsUnique)
{
    RoadLogistics logistics;
    auto          t1 = logistics.createTransport();
    auto          t2 = logistics.createTransport();
    EXPECT_NE(t1.get(), t2.get());
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
